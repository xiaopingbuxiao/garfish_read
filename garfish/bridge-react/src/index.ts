export { reactBridge } from './reactBridge';
export type { PropsInfo } from './types';
