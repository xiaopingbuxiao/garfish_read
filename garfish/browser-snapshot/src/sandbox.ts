import { PatchEvent } from './patchers/event';
import { PatchStyle } from './patchers/style';
import { PatchHistory } from './patchers/history';
import { PatchInterval } from './patchers/interval';
import { PatchGlobalVal } from './patchers/variable';
import { PatchWebpackJsonp } from './patchers/webpackjsonp';

export class Sandbox {
  public type = 'snapshot';
  public isRunning: Boolean = false;
  private patchList: Array<
    PatchGlobalVal | PatchStyle | PatchInterval | PatchEvent | PatchWebpackJsonp
  > = [];

  constructor(
    public name: string,
    public protectVariable: Array<PropertyKey> = [],
    public targetToProtect: Window | Object = typeof window !== 'undefined'
      ? window
      : globalThis,
    private isInBrowser: Boolean = typeof window === 'undefined' ? false : true,
  ) {
    this.name = name;
    this.isInBrowser = isInBrowser;
    this.patchList.push(new PatchGlobalVal(targetToProtect, protectVariable));

    // 执行顺序是，全局环境变量先激活，全局环境变量后销毁
    if (this.isInBrowser) {
      this.patchList = [
        ...this.patchList,
        new PatchStyle(),
        new PatchEvent(),
        new PatchHistory(),
        new PatchInterval(),
        new PatchWebpackJsonp(),
      ];
    }
  }

  //  1.触发生命周期钩子，willActivate（将要激活）
  //  2.将当前组的其他沙盒disable，并触发switch生命周期
  //  3.将当前window对象属性进行缓存
  //  4.获取style节点，进行缓存
  //  5.恢复沙盒运行期间产生的副作用
  public activate() {
    if (this.isRunning) return;
    this.patchList.forEach((patch) => {
      patch.activate();
    });
    this.isRunning = true;
  }

  // 1.恢复沙盒启动期间变量变更的变量，记录变更记录
  // 2.恢复沙盒启动期间删除的变量，记录变更记录
  public deactivate(clearEffects: boolean = true) {
    if (!this.isRunning) return; // 最后销毁全局变量守护
    [...this.patchList].reverse().forEach((patch) => {
      patch.deactivate(clearEffects);
    });
    this.isRunning = false;
  }
}
