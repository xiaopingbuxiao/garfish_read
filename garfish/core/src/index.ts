export type { interfaces } from './interface';
export { Garfish as default } from './garfish';
