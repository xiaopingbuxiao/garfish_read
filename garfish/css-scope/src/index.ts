export { parse } from './cssParser';
export { stringify } from './cssStringify';
export { CssScopeOptions, GarfishCssScope } from './pluginify';