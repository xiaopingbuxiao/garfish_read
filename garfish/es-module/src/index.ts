export { Runtime as default } from './runtime';
export { GarfishEsModule } from './pluginify';
