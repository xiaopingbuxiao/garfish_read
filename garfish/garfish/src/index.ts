export type { interfaces } from '@garfish/core';
export { default as Garfish } from '@garfish/core';
export { GarfishInstance as default } from './instance';
export { defineCustomElements } from './customElement';
