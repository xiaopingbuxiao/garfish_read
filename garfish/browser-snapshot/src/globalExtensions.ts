import { Sandbox } from './sandbox';

declare module '@garfish/core' {
  export namespace interfaces {
    export interface Config {
      protectVariable?: PropertyKey[];
      insulationVariable?: PropertyKey[];
      sandbox?: SandboxConfig | false;
    }

    export interface App {
      snapshotSandbox?: Sandbox;
    }

    export interface Plugin {
      openBrowser?: boolean;
    }
  }
}
