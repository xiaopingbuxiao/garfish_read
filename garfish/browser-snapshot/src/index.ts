import { Sandbox } from './sandbox';
import { interfaces } from '@garfish/core';
import './globalExtensions';

export { Sandbox as default } from './sandbox';

export interface SandboxConfig {
  snapshot?: boolean;
  disableWith?: boolean;
  strictIsolation?: boolean;
}

interface BrowserConfig {
  open?: boolean;
  protectVariable?: PropertyKey[];
}

export function GarfishBrowserSnapshot(op?: BrowserConfig) {
  return function (Garfish: interfaces.Garfish): interfaces.Plugin {
    const options = {
      openBrowser: false,
      version: __VERSION__,
      name: 'browser-snapshot',

      afterLoad(appInfo, appInstance) {
        const config: BrowserConfig = op || { open: true };
        const sandboxConfig = appInfo.sandbox || Garfish?.options?.sandbox;
        if (
          sandboxConfig === false ||
          sandboxConfig.open === false ||
          sandboxConfig?.snapshot === false
        ) {
          config.open = false;
        }
        if (sandboxConfig) {
          config.protectVariable = [
            ...(Garfish?.options.protectVariable || []),
            ...(appInfo.protectVariable || []),
          ];
        }
        options.openBrowser = !!config.open;
        if (!config.open) return;
        if (appInstance) {
          // existing
          if (appInstance.snapshotSandbox) return;
          const sandbox = new Sandbox(appInfo.name, config.protectVariable);
          appInstance.snapshotSandbox = sandbox;
        }
      },

      beforeMount(appInfo, appInstance) {
        // existing
        if (!appInstance.snapshotSandbox) return;
        appInstance.snapshotSandbox.activate();
      },

      afterUnmount(appInfo, appInstance) {
        if (!appInstance.snapshotSandbox) return;
        appInstance.snapshotSandbox.deactivate();
      },
    };
    return options;
  };
}
