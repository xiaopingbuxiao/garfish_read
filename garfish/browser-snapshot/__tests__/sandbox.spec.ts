import { Sandbox } from '../src/sandbox';

describe('test sandbox ', () => {
  it('dom sandbox', () => {
    document.head.innerHTML = `
      <style>body{color: 'yellow';}</style>
    `;

    const obj: {
      name: string;
      age: string;
      some?: string;
    } = {
      name: 'zhoushaw',
      age: '24',
      some: 'test delete',
    };

    const sb = new Sandbox('app1', [], obj);

    const st = document.createElement('style');
    st.style.cssText = 'background: red;';
    document.head.appendChild(st);
    obj.name = 'wu';
    obj.age = '23';
    delete obj.some;

    sb.deactivate();

    const result = document.querySelectorAll('style');
    expect(result).toMatchSnapshot();
    expect(obj).toMatchSnapshot();
  });
});
