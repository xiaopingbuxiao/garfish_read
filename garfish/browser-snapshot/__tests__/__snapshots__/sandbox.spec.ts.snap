// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test sandbox  dom sandbox 1`] = `
NodeList [
  <style>
    body{color: 'yellow';}
  </style>,
  <style
    style="background: red;"
  />,
]
`;

exports[`test sandbox  dom sandbox 2`] = `
Object {
  "age": "23",
  "name": "wu",
}
`;
